include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+{
+public:
+    Knob(const std::string& flagName)
+        : mFlagName(flagName), mDisplayName(flagName), mIndent(0) {}
+    virtual ~Knob() = default;
+
+    std::string GetDisplayName() const { return mDisplayName; }
+    std::string GetFlagName() const { return mFlagName; }
+    std::string GetFlagHelp() const { return mFlagDesc; }
+    size_t      GetIndent() const { return mIndent; }
+
+    // Populate flag members
+    void SetDisplayName(std::string displayName) { mDisplayName = displayName; }
+    void SetFlagDesc(std::string flagDesc) { mFlagDesc = flagDesc; }
+    void SetIndent(size_t indent) { mIndent = indent; }
+
+    virtual void Draw() = 0;
+
+    // Flag description in usage message
+    virtual std::string GetFlagHelpText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual void UpdateFromFlag(const CliOptions& opts) = 0;
+
+    // Default values are the values when the application starts. The knobs can be reset to default by a button in the UI
+    virtual void ResetToDefault() = 0;
+
+protected:
+    std::string mDisplayName;
+    std::string mFlagName;
+    std::string mFlagDesc;
+    size_t      mIndent;
+};
+
+// KnobCheckbox will be displayed as a checkbox in the UI
+// Currently supports variants of type: bool
+template <typename T, std::enable_if_t<std::is_same_v<T, bool>, bool> = true>
+class KnobCheckbox final
+    : public Knob
+{
+public:
+    KnobCheckbox(const std::string& flagName, T defaultValue)
+        : Knob(flagName)
+    {
+        SetDefault(defaultValue);
+    }
+
+    void Draw() override
+    {
+        ImGui::Checkbox(mDisplayName.c_str(), &mValue);
+    }
+
+    void ResetToDefault() override { SetValue(mDefaultValue); }
+
+    // GetFlagHelpText format:
+    // --flag_name <true|false> : flag description
+    std::string GetFlagHelpText() const override
+    {
+        std::string flagHelpText = "--" + mFlagName + " <true|false>";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Updates mValue and mDefaultValue, expected flag format:
+    // --flag_name <true|false>
+    void UpdateFromFlag(const CliOptions& opts) override
+    {
+        SetDefault(opts.GetExtraOptionValueOrDefault(mFlagName, mValue));
+    }
+
+    T GetValue() const { return mValue; }
+
+    // Used for when mValue needs to be updated outside of UI
+    // Updates mValue
+    void SetValue(T newVal)
+    {
+        mValue = newVal;
+    }
+
+private:
+    // SetDefault will set mDefaultValue and update mValue
+    void SetDefault(T newVal)
+    {
+        mDefaultValue = newVal;
+        ResetToDefault();
+    }
+
+private:
+    T mValue;
+    T mDefaultValue;
+};
+
+// KnobSlider will be displayed as a slider in the UI
+// ImGui sliders can also become input boxes with ctrl + right click
+// Currently supports variants of type: int
+template <typename T, std::enable_if_t<std::is_same_v<T, int>, bool> = true>
```

Comment:
Also use static assert...

Reply:
Why use static assert here?
For the current PR KnobSlider is only accepting `int` to reduce the PR length, but slider will definitely require `float` too.
Seems like enable_if construct is clearly intended for such cases no?

Reply:
Usually enable_if is used for type trait dependent specialization. It creates cryptic error message, since the compiler can't find the appropriate type. For example MSVC:

enable_if
```
error C2972: 'KnobSlider': template parameter 'unnamed-parameter': the type of non-type argument is invalid
note: see declaration of 'KnobSlider'
error C2672: 'CreateKnob': no matching overloaded function found
note: could be 'T CreateKnob(void)'
note: 'CreateKnob': invalid template argument for 'T', missing template argument list on class template 'KnobSlider'
```

static_assert
```
error C2338: static_assert failed: 'Slider support int only'
note: see reference to class template instantiation 'KnobSlider<float>' being compiled
```

Reply:
Template specialization also works here if we just want to implement for `int`:

```cpp
template<typename T>
class KnobSlider;

template<>
class KnobSlider<int> {
  typedef int ValueType;
  //...
  ValueType GetValue() const { return mValue; }
  //...
}
```
that also gives better error messages (e.g. with `CreateKnob<KnobSlider<float>>()`):
```
error C2027: use of undefined type 'KnobSlider<float>'
note: see declaration of 'KnobSlider<float>'
```